libs/qCC_io/ShpFilter.cpp

```diff
@@ -651,6 +718,405 @@ static CC_FILE_ERROR LoadPolyline(QDataStream &shpStream,
+	size_t numPoints = points.size();
+
+
+	for (int32_t i = 0; i < numParts; ++i)
+	{
+		if (!isValidEsriPartType(partTypes[i]))
+		{
+			ccLog::Warning("[SHP] Multipatch part %d has an invalid part type (%d)", i, partTypes[i]);
+			continue;
+		}
+		auto type = static_cast<ESRI_PART_TYPE>(partTypes[i]);
+
+		const int32_t &firstIndex = startIndexes[i];
+		const int32_t &lastIndex = static_cast<const int32_t &>((i + 1 < numParts ? startIndexes[i + 1] : numPoints) -
+		                                                        1);
+		const int32_t vertCount = lastIndex - firstIndex + 1;
+
+		switch (type)
+		{
+			case ESRI_PART_TYPE::TRIANGLE_STRIP:
+			{
+				ccMesh *mesh = createMesh(points, scalarValues, firstIndex, lastIndex);
+				for (int32_t j(2); j < vertCount; ++j)
+				{
+					mesh->addTriangle(j - 2, j - 1, j);
+				}
+				container.addChild(mesh);
+				break;
+			}
+			case ESRI_PART_TYPE::TRIANGLE_FAN:
+			{
+				ccMesh *mesh = createMesh(points, scalarValues, firstIndex, lastIndex);
+				for (int32_t j(2); j < vertCount; ++j)
+				{
+					mesh->addTriangle(0, j - 1, j);
+				}
+				container.addChild(mesh);
+				break;
+			}
+			default:
+				ccLog::Print(QString("[SHP] Cannot handle Patch of type: %1").arg(ToString(type)));
+				return CC_FERR_BAD_ENTITY_TYPE;
+		}
+	}
+	return CC_FERR_NO_ERROR;
+}
+
+
+static CC_FILE_ERROR LoadMultiPatch(QDataStream &shpStream,
+                                    ccHObject &container,
+                                    CCVector3d Pshift)
+{
+	// skip record bbox
+	shpStream.skipRawData(4 * sizeof(double));
+
+	int32_t numParts;
+	int32_t numPoints;
+	shpStream >> numParts >> numPoints;
+
+	std::vector<int32_t> startIndexes = readParts(shpStream, numParts);
+	if (startIndexes.empty())
+	{
+		return CC_FERR_NOT_ENOUGH_MEMORY;
+	}
+
+	std::vector<int32_t> partTypes = readParts(shpStream, numParts);
+	if (partTypes.empty())
+	{
+		return CC_FERR_NOT_ENOUGH_MEMORY;
+	}
+
+
+	std::vector<CCVector3> points = readPoints(shpStream, numPoints, Pshift);
+	if (points.empty())
+	{
+		return CC_FERR_NOT_ENOUGH_MEMORY;
+	}
+
+	//Z boundaries
+	shpStream.skipRawData(2 * sizeof(double));
+
+	//Z coordinates (an array of length NumPoints)
+	for (int32_t i = 0; i < numPoints; ++i)
+	{
+		double z;
+		shpStream >> z;
+		points[i].z = static_cast<PointCoordinateType>(z + Pshift.z);
+	}
+
+	std::vector<ScalarType> scalarValues = readMeasures(shpStream, numPoints);
+	return buildPatches(container, startIndexes, partTypes, points, scalarValues);
+}
+
+//! Saves the cloud to the shape file according to the specification
+/**
+ *
+ * @param stream Output stream of the shapefile
+ * @param cloud The cloud to save (pointcloud or vertices)
+ * @param bbMing Min point of the cloud
+ * @param bbMaxg Mxx point of the cloud
+ */
+static void save3DCloud(QDataStream &stream, ccGenericPointCloud *cloud, const CCVector3d &bbMing, const CCVector3d &bbMaxg)
+{
+	unsigned numPoints = cloud->size();
```

Comment:
const?

Reply:
const for the `ccGenericPointCloud` right ?

Reply:
I'd meant `numPoints` (I think it is const?), but if `cloud` can be too, that's good as well.